### ISSUE TITLE
Dill fails to pickle class instance defined in function scope with super() call in __init__.

### ISSUE THREAD

**Issue by asaitgalin** [uqfoundation/dill#209]
Here is the code demonstrating problem:
```
import dill

def main():
    class A(object): pass
    class B(A):
        def __init__(self): super(B, self).__init__()
        def do(self): return 5

    dill.dumps(B())

if __name__ == "__main__":
    main()
```

which fails with `RuntimeError: maximum recursion depth exceeded while calling a Python object`.

I found out that `B` class is added to `__init__` closure (it seems logical, since B type is defined only in `main` function and proxy object returned by `super` may need this type). So, there is a cycle in dump process:
`B -> B dict -> __init__ function -> __init__ function closure -> B`

This can be fixed in code above, for example, using `A.__init__(self)` instead of `super()` call.

I suppose this can be fixed in dill by dropping class from method closure if method is defined in the class with special bool flag that function should contain this class in closure. On unpickling the original class can be added back to closure depending on this flag. But I don't feel to have enough expertise in the question :)

**Comment by mmckerns** [comment]
@asaitgalin: I think this is a duplicate of #75.  I'm working on it presently. There is a workaround for user defined classes. If this is a duplicate, you can mark it as duplicate and close the issue.

**Comment by mmckerns** [comment]
I've closed #75.  I think the intent of this ticket was the same as that issue... and indeed, if you define class `A` and class `B` outside of `main`, the code now works.  However, with classes `A` and `B` as is (defined inside of `main`), the code still fails.  Not with a recursion error, but with:
```
pickle.PicklingError: Can't pickle <class '__main__.B'>: it's not found as __main__.B
```

**Comment by mmckerns** [comment]
I'm pretty sure what's left is a duplicate of #56.  I'm closing this. Reopen it if you disagree.
